File: scripts/auditoria_maestra.py
import subprocess
import sys
import json
import os
from datetime import datetime

# Configuración
TMP_DIR = ".tmp"
REPORT_JSON = os.path.join(TMP_DIR, "reporte_auditoria_final.json")
REPORT_MD = os.path.join(TMP_DIR, "reporte_auditoria_final.md")

if not os.path.exists(TMP_DIR):
    os.makedirs(TMP_DIR)

audit_results = {
    "timestamp": datetime.now().isoformat(),
    "steps": []
}

def run_step(name, command, critical=True):
    """Ejecuta un paso de auditoría y registra el resultado."""
    print(f"\n--- Ejecutando: {name} ---")
    start_time = datetime.now()
    try:
        # Ejecutar comando
        result = subprocess.run(command, shell=True, capture_output=True, text=True)
        duration = (datetime.now() - start_time).total_seconds()
        
        success = result.returncode == 0
        output = result.stdout + "\n" + result.stderr
        
        step_result = {
            "name": name,
            "success": success,
            "duration": duration,
            "command": command,
            "output": output[:2000] # Truncar salida para el JSON
        }
        
        audit_results["steps"].append(step_result)
        
        if success:
            print(f"✅ {name}: PASSED ({duration:.2f}s)")
        else:
            print(f"❌ {name}: FAILED ({duration:.2f}s)")
            print(f"Output:\n{output[:500]}...") # Mostrar error
            
            if critical:
                print("⛔ Error crítico detectado. Abortando auditoría.")
                save_report()
                sys.exit(1)
                
        return success

    except Exception as e:
        print(f"❌ {name}: ERROR ({str(e)})")
        audit_results["steps"].append({
            "name": name,
            "success": False,
            "error": str(e)
        })
        if critical:
            save_report()
            sys.exit(1)
        return False

def save_report():
    """Guarda el reporte en JSON y Markdown."""
    # Guardar JSON
    with open(REPORT_JSON, "w", encoding="utf-8") as f:
        json.dump(audit_results, f, indent=2)
        
    # Generar Markdown
    md_content = f"# Reporte de Auditoría Final MVP\n\n"
    md_content += f"**Fecha:** {audit_results['timestamp']}\n\n"
    md_content += "| Paso | Estado | Duración | Detalle |\n"
    md_content += "|---|---|---|---|\n"
    
    all_passed = True
    
    for step in audit_results["steps"]:
        status_icon = "✅" if step["success"] else "❌"
        if not step["success"]:
            all_passed = False
        md_content += f"| {step['name']} | {status_icon} | {step.get('duration', 0):.2f}s | Ver JSON para detalles |\n"
        
    md_content += "\n## Conclusión\n"
    if all_passed:
        md_content += "🎉 **LA APP ESTÁ LISTA PARA MVP (Técnicamente)**. Todos los checks pasaron.\n"
    else:
        md_content += "⚠️ **SE ENCONTRARON ERRORES**. Revisar log detallado.\n"

    with open(REPORT_MD, "w", encoding="utf-8") as f:
        f.write(md_content)
        
    print(f"\n📄 Reporte guardado en: {REPORT_MD}")

def main():
    print("🚀 Iniciando Auditoría Maestra MVP...\n")
    
    # 1. Linting
    run_step("Linting (Análisis Estático)", "npm run lint", critical=False)
    
    # 2. Type Check / Build
    run_step("Build (TypeScript Check)", "npm run build", critical=True)
    
    # 3. Pruebas Unitarias
    run_step("Unit Tests (Vitest)", "npm run test", critical=True)
    
    # 4. Seguridad (Script Python)
    # Nota: Asumiendo que python está en path como 'python' o 'python3'
    python_cmd = "python" if sys.platform == "win32" else "python3"
    run_step("Seguridad (RLS Check)", f"{python_cmd} scripts/verificar_aislamiento_datos.py", critical=True)
    
    # 5. Guardar
    save_report()

if __name__ == "__main__":
    main()

File: scripts/auditoria_maestra.js
import { exec } from 'child_process';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import util from 'util';

// Promisify exec
const execAsync = util.promisify(exec);

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configuración
const TMP_DIR = path.resolve(__dirname, '../.tmp');
const REPORT_JSON = path.join(TMP_DIR, "reporte_auditoria_final.json");
const REPORT_MD = path.join(TMP_DIR, "reporte_auditoria_final.md");

if (!fs.existsSync(TMP_DIR)) {
    fs.mkdirSync(TMP_DIR, { recursive: true });
}

const auditResults = {
    timestamp: new Date().toISOString(),
    steps: []
};

async function runStep(name, command, critical = true) {
    console.log(`\n--- Ejecutando: ${name} ---`);
    const startTime = Date.now();
    let success = false;
    let output = "";

    try {
        const { stdout, stderr } = await execAsync(command);
        output = stdout + "\n" + stderr;
        success = true; // Si exec no lanza error, el exit code fue 0
        console.log(`✅ ${name}: PASSED`);
    } catch (error) {
        success = false;
        output = error.stdout + "\n" + error.stderr;
        console.log(`❌ ${name}: FAILED`);
        console.log(`Output:\n${output.substring(0, 500)}...`);
    }

    const duration = (Date.now() - startTime) / 1000;

    auditResults.steps.push({
        name,
        success,
        duration,
        command,
        output: output.substring(0, 2000) // Truncar
    });

    if (!success && critical) {
        console.log("⛔ Error crítico detectado. Abortando auditoría.");
        saveReport();
        process.exit(1);
    }

    return success;
}

function saveReport() {
    // Guardar JSON
    fs.writeFileSync(REPORT_JSON, JSON.stringify(auditResults, null, 2), 'utf-8');

    // Generar Markdown
    let mdContent = `# Reporte de Auditoría Final MVP\n\n`;
    mdContent += `**Fecha:** ${auditResults.timestamp}\n\n`;
    mdContent += `| Paso | Estado | Duración | Detalle |\n`;
    mdContent += `|---|---|---|---|\n`;

    let allPassed = true;

    auditResults.steps.forEach(step => {
        const statusIcon = step.success ? "✅" : "❌";
        if (!step.success) allPassed = false;
        mdContent += `| ${step.name} | ${statusIcon} | ${step.duration.toFixed(2)}s | Ver JSON para detalles |\n`;
    });

    mdContent += `\n## Conclusión\n`;
    if (allPassed) {
        mdContent += `🎉 **LA APP ESTÁ LISTA PARA MVP (Técnicamente)**. Todos los checks pasaron.\n`;
    } else {
        mdContent += `⚠️ **SE ENCONTRARON ERRORES**. Revisar log detallado.\n`;
    }

    fs.writeFileSync(REPORT_MD, mdContent, 'utf-8');
    console.log(`\n📄 Reporte guardado en: ${REPORT_MD}`);
}

async function main() {
    console.log("🚀 Iniciando Auditoría Maestra MVP (Node.js Version)...\n");

    try {
        // 1. Linting
        await runStep("Linting (Análisis Estático)", "npm run lint", false);

        // 2. Type Check / Build
        await runStep("Build (TypeScript Check)", "npm run build", true);

        // 3. Pruebas Unitarias
        await runStep("Unit Tests (Vitest)", "npm run test", true);

        // 4. Seguridad (Script JS)
        // Usamos node para ejecutar el script hermano
        const scriptPath = path.join(__dirname, 'verificar_aislamiento_datos.js');
        // Necesitamos asegurar que node pueda ejecutar modulos ES, así que usamos node directamente
        await runStep("Seguridad (RLS Check)", `node "${scriptPath}"`, true);

    } catch (err) {
        console.error("Error fatal en el orquestador:", err);
    } finally {
        // 5. Guardar
        saveReport();
    }
}

main();
